Update Testerina API Docs (#30753)
ballerina-platform/ballerina-lang: misc/testerina/modules/testerina-core/src/main/ballerina/Module.md

## Module overview

This module facilitates developers to write automation tests for Ballerina code in a simple manner. 
It provides a number of capabilities such as configuring setup and cleanup steps in different 
levels, ordering and grouping of tests, providing value-sets to tests and independence from external
functions and endpoints via mocking capabilities.

## Annotations
A Ballerina testsuite can be implemented using a set of annotations. The available annotations 
enable executing instructions before and after the testsuite or a single test, organize a set of 
tests into a group, define data-driven tests, specify an order of execution, disable tests and 
mocking.

### Test Config 

The following example shows a simple testsuite.
```ballerina
import ballerina/test;

// Test function
@test:Config {}
function testFunction() {
    test:assertTrue(true, msg = "Failed!");
}
```

> The test config annotation can be used to set the test execution order by specifying the before and 

Reviewer: It should be The `before` and `after` attributes, right? To be consistent with the details you've provided about other attributes.

Author: Made it consistent with the other details

## Module overview

This module facilitates developers to write automation tests for Ballerina code in a simple manner. 

> It provides a number of capabilities such as configuring setup and cleanup steps in different 

Reviewer 2: ```suggestion
It provides a number of capabilities such as the configuring setup and cleanup steps in different 
```

> levels, ordering and grouping of tests, providing value-sets to tests and independence from external

Reviewer 2: ```suggestion
levels, ordering and grouping of tests, providing value-sets to tests, and independence from the external
```

functions and endpoints via mocking capabilities.

## Annotations
A Ballerina testsuite can be implemented using a set of annotations. The available annotations 
enable executing instructions before and after the testsuite or a single test, organize a set of 
tests into a group, define data-driven tests, specify an order of execution, disable tests and 
mocking.

### Test Config 

The following example shows a simple testsuite.
```ballerina
import ballerina/test;

// Test function
@test:Config {}
function testFunction() {
    test:assertTrue(true, msg = "Failed!");
}
```

The test config annotation can be used to set the test execution order by specifying the before and 
after functions.

```ballerina
import ballerina/io;
import ballerina/test;

function beforeFunc() {
    io:println("I'm the before function!");
}

@test:Config {
    before: beforeFunc,
    after: afterFunc
}
function testFunction() {
    io:println("I'm in test function!");
    test:assertTrue(true, msg = "Failed!");
}

function afterFunc() {
    io:println("I'm the after function!");
}
```

The `dependsOn` attribute can also be used to set the test execution order by specifying the list of

> functions that the test function dependsOn.

Reviewer 2: ```suggestion
functions that the test function depends on.
```

// Test function
@test:Config {

>     before: "beforeFunc",

Reviewer: Since we say that it's a simple test suite, shall we remove all these annotations and attributes and make it the simplest test suite with only one test case? 
Then, may be we can add and describe the following capabilities.

1. Setup and tear down pre-requisites and post actions for a test suite, test group, per each test or for a single test. We can include a sample for beforeSuite and afterSuite and mention that similarly this can be achieved for other cases as well.

2. Provide data sets to run the tests against using data provider

3. Ensure test execution order using 'dependsOn'

WDYT?

Reviewer: This https://github.com/ballerina-platform/ballerina-distribution/blob/master/doc-guidelines.md#ballerina-api-documentation-guidelines says,

> Module-level documentation page needs to introduce the general concept of the API and the main APIs that are used.

}
```

### Before and After Suite

The `BeforeSuite` annotation is used to execute a particular function before the test suite is

> executed. This can be used to setup prerequisites before executing the test suite. 

Reviewer 2: ```suggestion
executed. This can be used to set up the prerequisites before executing the test suite. 
```

```ballerina
@test:BeforeSuite
function beforeSuit() {
    // initialize or execute pre-requisites
}

```

The `AfterSuite` annotation is used to execute a particular function after the test suite is

> executed. This is used to tear-down prerequisites or execute post actions after executing the test

Reviewer 2: ```suggestion
executed. This is used to tear down the prerequisites or execute the post actions after executing the test
```

suite.

```ballerina
@test:AfterSuite
function afterSuit() {
    // tear-down
}

```

### Before and After Groups

The `BeforeGroups` and annotation can be used to execute the function before the first test function

> belonging the specified groups.

Reviewer 2: ```suggestion
belonging to the specified groups.
```

```ballerina
import ballerina/io;
import ballerina/test;

@test:BeforeGroups { value:["g1"] }
function beforeGroupsFunc() {
    io:println("I'm the before groups function!");
}

@test:Config { groups: ["g1"]}
function testFunction1() {
    io:println("I'm in test function 1!");
    test:assertTrue(true, msg = "Failed!");
}
```

> The `AfterGroups` and annotation can be used to execute the function after the last test function

Reviewer 2: ```suggestion
The `AfterGroups` annotation can be used to execute the function after the last test function
```

> belonging the specified groups.

Reviewer 2: ```suggestion
belonging to the specified groups.
```

```ballerina
import ballerina/io;
import ballerina/test;

@test:AfterGroups { value:["g1"] }
function afterGroupsFunc() {
    io:println("I'm the after groups function!");
}

@test:Config { groups: ["g1"]}
function testFunction1() {
    io:println("I'm in test function 1!");
    test:assertTrue(true, msg = "Failed!");
}
```

## Assertions
This module provides a number of assertions in order to verify the expected behaviour of a piece of code. 
These assertions can be used to decide if the test is passing or failing based on the condition.

> Following sample shows how to use assertions in Testerina.

Reviewer 2: ```suggestion
The following sample shows how to use assertions in Testerina.
```

The following example shows different ways of stubbing a function call.

```ballerina

> import ballerina/test;

Author: Do we need to explicitly include `import ballerina/test;` in the samples? its a bit inconsistent throughout the document. 

WDYT?

Author: Can these code snippets be considered examples that are meant to be copied to a project or just references?

Reviewer: Let's add the imports wherever applicable.
These are considered to be examples IMO.

Author: Added imports where neccessary

_main.bal_
```ballerina
http:Client clientEndpoint = check new ("http://petstore.com");

function getPet(string petId) returns Pet | error {

>   http:Response|error result = check clientEndpoint->get("/pets?id="+petId);

Reviewer: ```suggestion
  http:Response result = check clientEndpoint->get("/pets?id="+petId);
```

>   Pet pet = constructPetObj(result); 

Author: This section of `Object Mocking` has been updated to the current standard, however, the sample itself is incomplete. One cant simply copy this example and run as it would require some more object definitions and other functions to be defined

Do we proceed with the existing sample? Since we have examples elsewhere, shall we keep it as is?

Reviewer: Shall we add a simple `constructPetObj` function and also any other missing functions/definitions? Let's use a complete sample. If you feel it's getting large, let's add another simple example. WDYT?

Author: I changed the example a bit to exclude this `constructPetObj`. I have no clue what the original function intended to do but the current example simply demonstrates how it works. 

We need to review these examples specifically.

  return pet;
}
```

_test.bal_
```ballerina
// Mock object definition for http:Client object
public client class MockHttpClient {
    remote isolated function get(@untainted string path, map<string|string[]>? headers = (),

>     http:TargetType targetType = http:Response) returns http:Response | http:ClientError {

Reviewer: ```suggestion
    http:TargetType targetType = http:Response) returns http:Response | http:PayloadType | http:ClientError {
```

> The annotation `@test:Mock {}` is used to declare a `MockFunction` object, with details of the name of the function to be mocked, as well as the module name if an import function is being mocked. The module name value of the annotation is optional if the function being mocked is not an import function.

Reviewer 2: ```suggestion
The annotation `@test:Mock {}` is used to declare a `MockFunction` object with details of the name of the function to be mocked as well as the module name if an import function is being mocked. The module name value of the annotation is optional if the function being mocked is not an import function.
```

_main.bal_
```ballerina
import ballerina/http;

http:Client clientEndpoint = check new ("http://petstore.com");

function getPet(string petId) returns http:Response | error {
  http:Response result = check clientEndpoint->get("/pets?id="+petId);
  return result;
}

function deletePet(string petId) returns error? {
    http:Response res = check clientEndpoint->delete("/pets/delete?id="+petId);
}
```

_test.bal_
```ballerina
import ballerina/test;
import ballerina/http;

> // Mock object definition for http:Client object

Reviewer 2: ```suggestion
// Mock object definition for the `http:Client` object
```

public client class MockHttpClient {
    remote isolated function get(@untainted string path, map<string|string[]>? headers = (),
    http:TargetType targetType = http:Response) returns http:Response | http:PayloadType | http:ClientError {
        http:Response response = new;
        response.statusCode = 500;
        return response;
    }
}

@test:Config {}
function testTestDouble() returns error? {
    //mock object that would act as the test double to the clientEndpoint
    clientEndpoint = test:mock(http:Client, new MockHttpClient());
    http:Response res = check getPet("D123");
    test:assertEquals(res.statusCode, 500);
}
```

#### Stubbing member functions and variables of an object

> The member functions and variables are stubbed to return a specific value or to do nothing when invoked. Using the test module, a default mock object of the specified type. The default action of any member function/variable is to panic upon invocation/retrieval.  After mock object creation, the required functions and variables of the default mock object should be stubbed to return a value or to do nothing when called.

Reviewer 2: ```suggestion
The member functions and variables are stubbed to return a specific value or to do nothing when invoked. Using the test module, a default mock object of the specified type. The default action of any member function/variable is to panic upon invocation/retrieval.  After mock object creation, the required functions and variables of the default mock object should be stubbed to return a value or to do nothing when called.
```
This sentence is not complete: "Using the test module, a default mock object of the specified type."

##### Samples

> Example

Reviewer: We can remove this right?

Author: Removed stray example

> Following example demonstrates how to stub a member function to return a specific value.

Reviewer 2: ```suggestion
The following example demonstrates how to stub a member function to return a specific value.
```

```ballerina

@test:Config {}
function testThenReturn() returns error? {
   clientEndpoint = test:mock(http:Client);

    http:Response mockResponse = new;
    mockResponse.statusCode = 300;

    // stubbing to return the specified value 
    test:prepare(clientEndpoint).when("get").thenReturn(mockResponse);

    http:Response res = check getPet("D123");
    test:assertEquals(res.statusCode, 300);
}
```

> Following example demonstrates how to stub a member function to return different values on subsequent calls.

Reviewer 2: ```suggestion
The following example demonstrates how to stub a member function to return different values on the subsequent calls.
```

```ballerina
@test:Config {}
function testThenReturnSequence() returns error? {
     clientEndpoint = test:mock(http:Client);

    http:Response mockResponse1 = new;
    mockResponse1.statusCode = 400;

    http:Response mockResponse2 = new;
    mockResponse2.statusCode = 401;
    

>     // stubbing to return different values for each function call

Reviewer 2: ```suggestion
    // Stubbing to return different values for each function call.
```

    test:prepare(clientEndpoint).when("get").thenReturnSequence(mockResponse1, mockResponse2);

    http:Response res = check getPet("D123");
    test:assertEquals(res.statusCode, 400);
    res = check getPet("D123");
    test:assertEquals(res.statusCode, 401);
}
```

> Following example demonstrates how to stub a member function to do nothing when called.

Reviewer 2: ```suggestion
The following example demonstrates how to stub a member function to do nothing when called.
```

```ballerina
@test:Config {}
function testDoNothing() returns error? {
    clientEndpoint = test:mock(http:Client);
    

>     // stubbing to return different values for each function call

Reviewer 2: ```suggestion
    // Stubbing to return different values for each function call.
```
